src/scrub.S

```diff
@@ -42,81 +81,37 @@ __metal_before_start:
 
     la      t0, __metal_boot_hart
     csrr    a5, mhartid
+
     /* Disable machine interrupts to be safe */
     li      a3, 8
     csrc    mstatus, a3
-    bne     a5, t0, wait_scrub
-
-    /* Zero out itim memory. */
-    .weak metal_itim_0_memory_start
-    .weak metal_itim_0_memory_end
-    la      t1, metal_itim_0_memory_start
-    la      t2, metal_itim_0_memory_end
-    beq     t1, t2, dtim_scrub
-    jal     _metal_memory_scrub
-    .weak metal_itim_1_memory_start
-    .weak metal_itim_1_memory_end
-    la      t1, metal_itim_1_memory_start
-    la      t2, metal_itim_1_memory_end
-    beq     t1, t2, dtim_scrub
-    jal     _metal_memory_scrub
-
-dtim_scrub:
-    /* Zero out dtim memory. */
-    .weak metal_dtim_0_memory_start
-    .weak metal_dtim_0_memory_end
-    la      t1, metal_dtim_0_memory_start
-    la      t2, metal_dtim_0_memory_end
-    beq     t1, t2, ils_scrub
-    jal     _metal_memory_scrub
-
-ils_scrub:
-    /* Zero out ils memory. */
-    .weak metal_ils_0_memory_start
-    .weak metal_ils_0_memory_end
-    la      t1, metal_ils_0_memory_start
-    la      t2, metal_ils_0_memory_end
-    beq     t1, t2, dls_scrub
-    jal     _metal_memory_scrub
-
-dls_scrub:
-    /* Zero out dls memory. */
-    .weak metal_dls_0_memory_start
-    .weak metal_dls_0_memory_end
-    la      t1, metal_dls_0_memory_start
-    la      t2, metal_dls_0_memory_end
-    beq     t1, t2, sram_scrub
-    jal     _metal_memory_scrub
 
-sram_scrub:
-    /* Zero out sram memory. */
-    .weak metal_sys_sram_0_memory_start
-    .weak metal_sys_sram_0_memory_end
-    la      t1, metal_sys_sram_0_memory_start
-    la      t2, metal_sys_sram_0_memory_end
-    beq     t1, t2, memory_scrub
-    jal     _metal_memory_scrub
-
-memory_scrub:
-    /* Zero out main memory. */
-    .weak metal_memory_0_memory_start
-    .weak metal_memory_0_memory_end
-    la      t1, metal_memory_0_memory_start
-    la      t2, metal_memory_0_memory_end
-    beq     t1, t2, done_scrub
-    jal     _metal_memory_scrub
-
-done_scrub:
-    lui     a4, 0x2000
-    li      a5,1
-    sw      a5,0(a4)
-    fence   w,rw
-    j skip_scrub
+    /* Zero out per hart stack */
+    mv      t1, sp
+    la      t2, __stack_size
+    add     t2, t2, sp
+    beq     t1, t2, 1f
+    jal     __metal_memory_scrub
+1:
+    bne     a5, t0, skip_scrub
 
-wait_scrub:
-    lui     a4, 0x2000
-    lw      a5, 0(a4)
-    beqz    a5, wait_scrub
+    /* Zero out data segment */
+    la      t1, metal_segment_data_target_start
+    la      t2, metal_segment_data_target_end
+    beq     t1, t2, 1f
+    jal     __metal_memory_scrub
+1:
+    /* Zero out itim memory */
+    la      t1, metal_segment_itim_target_start
+    la      t2, metal_segment_itim_target_end
+    beq     t1, t2, 1f
+    jal     __metal_memory_scrub
+1:
+    /* Zero out heap memory */
```

Comment:
I'm concerned that the heap may be large on some systems causing a long delay at this point in the program execution. That is why I suggested delaying this part of the memory scrub until sbrk is invoked. If metal_memory_scrub is careful about setting the system up to avoid exceptions, that should be fairly easy to manage.

Reply:
I added it here, since current default size of heap in fe-sdk wasn't large enough to cause much of an impact on start up times. Large sizes will lead to concern.
Updated _sbrk() with a call to metal_memory_scrub(), to scrub off heap as needed. This ultimately, limits 'malloc' to work only on m-mode [acceptable constraint ?]. beu interrupts cannot be delegated to s/u mode.